helm-chart/build.py

```diff
@@ -2,149 +2,193 @@
 
-def publish_pages():
-    version = last_git_modified('.')
+    chart['version'] = version
+
+    with open(chart_file, 'w') as f:
+        rt_yaml.dump(chart, f)
+
+
+def publish_pages(name, paths, git_repo, published_repo):
+    """publish helm chart index to github pages"""
+    version = last_modified_commit(*paths)
+    checkout_dir = '{}-{}'.format(name, version)
     subprocess.check_call([
         'git', 'clone', '--no-checkout',
-        'git@github.com:jupyterhub/helm-chart', 'gh-pages'],
-        env=dict(os.environ, GIT_SSH_COMMAND=f'ssh -i {HELM_CHART_DEPLOY_KEY_NAME}')
+        'git@github.com:{}'.format(git_repo), checkout_dir],
     )
-    subprocess.check_call(['git', 'checkout', 'gh-pages'], cwd='gh-pages')
-    subprocess.check_call(['helm', 'repo', 'add', 'jupyterhub', 'https://jupyterhub.github.io/helm-chart/'])
-    subprocess.check_call(['helm', 'repo', 'update'])
-    subprocess.check_call([
-        'helm', 'package', '--dependency-update', CHARTPATH,
-        '--destination', 'gh-pages/'
-    ])
-    subprocess.check_call([
-        'helm', 'repo', 'index', '.',
-        '--url', 'https://jupyterhub.github.io/helm-chart'
-    ], cwd='gh-pages')
-    subprocess.check_call(['git', 'add', '.'], cwd='gh-pages')
+    subprocess.check_call(['git', 'checkout', 'gh-pages'], cwd=checkout_dir)
+
+    # package the latest version into a temporary directory
+    # and run helm repo index with --merge to update index.yaml
+    # without refreshing all of the timestamps
+    with TemporaryDirectory() as td:
+        subprocess.check_call([
+            'helm', 'package', name,
+            '--destination', td + '/',
+        ])
+
+        subprocess.check_call([
+            'helm', 'repo', 'index', td,
+            '--url', published_repo,
+            '--merge', os.path.join(checkout_dir, 'index.yaml'),
+        ])
+
+        # equivalent to `cp td/* checkout/`
+        # copies new helm chart and updated index.yaml
+        for f in os.listdir(td):
+            shutil.copy2(
+                os.path.join(td, f),
+                os.path.join(checkout_dir, f)
+            )
+    subprocess.check_call(['git', 'add', '.'], cwd=checkout_dir)
     subprocess.check_call([
         'git',
         'commit',
-        '-m', '[binderhub] Automatic update for commit {}'.format(version)
-    ], cwd='gh-pages')
+        '-m', '[{}] Automatic update for commit {}'.format(name, version)
+    ], cwd=checkout_dir)
     subprocess.check_call(
         ['git', 'push', 'origin', 'gh-pages'],
-        cwd='gh-pages',
-        env=dict(os.environ, GIT_SSH_COMMAND=f'ssh -i ../{HELM_CHART_DEPLOY_KEY_NAME}')
+        cwd=checkout_dir,
     )
 
 
 def main():
-    argparser = argparse.ArgumentParser()
-    argparser.add_argument(
-        '--image-prefix',
-        default='jupyterhub/k8s-'
-    )
-    subparsers = argparser.add_subparsers(dest='action')
+    with open('chartpress.yaml') as f:
+        safe_yaml = YAML(typ='safe')
```

Comment:
type?

Reply:
I simplified the yaml initialization in the latest commit with a comment: `typ` (chosen presumably to avoid colliding with builtin `type`) is used to signal which type of yaml loader is used. We use `typ='rt'` for the roundtrip loader (which is also 'safe') since we are editing files.